psi4/src/psi4/libpe/psipe.cc

```diff
@@ -0,0 +1,259 @@
+ */
+#ifdef USING_cppe
+
+#include <iomanip>
+
+#include <armadillo>
+
+#include "psi4/psi4-dec.h"
+
+#include "psi4/libmints/basisset.h"
+#include "psi4/libmints/dimension.h"
+#include "psi4/libmints/integral.h"
+#include "psi4/libmints/efpmultipolepotential.h"
+#include "psi4/libmints/matrix.h"
+#include "psi4/libmints/molecule.h"
+#include "psi4/libmints/petitelist.h"
+#include "psi4/libmints/potentialint.h"
+#include "psi4/libmints/vector.h"
+#include "psi4/libpsi4util/PsiOutStream.h"
+
+// #include <cppe/libcppe.hh>
+#include <cppe/utils/potfile_reader.hh>
+#include <cppe/utils/pot_manipulation.hh>
+#include <cppe/core/multipole_expansion.hh>
+
+
+#include "psipe.h"
+
+// TODO: implement output stream option in CPPE for printing
+
+namespace psi {
+    
+libcppe::Molecule make_molecule(std::shared_ptr<Molecule> molecule) {
+    libcppe::Molecule mol;
+    Matrix geom = molecule->geometry();
+    for (int i = 0; i < molecule->natom(); ++i) {
+        libcppe::Atom a(static_cast<int>(molecule->Z(i)),
+                                         geom.get(i, 0),
+                                         geom.get(i, 1),
+                                         geom.get(i, 2));
+        mol.push_back(a);
+    }
+    return mol;
+}
+    
+// TODO: generate these factors!
+namespace {
+    std::vector<double> mult_coeffs = {1.0, // q
+                1.0, 1.0, 1.0, // mu
+                0.5, 0.5, 0.5, 1., 1., 1. // theta
+            };
+                
+    std::vector<int> from_alphabetical_to_psi4 = {
+            0,
+            1, 2, 3,
+            4, 7, 9, 5, 6, 8 // 0, 3, 5, 1, 2, 4   += 4
+    };
+} // unnamed namespace
+
+  
+PeState::PeState(libcppe::PeOptions options, std::shared_ptr<BasisSet> basisset) :
+      basisset_(basisset),
+      cppe_state_(libcppe::CppeState(options, make_molecule(basisset_->molecule()), *(outfile->stream()))),
+      int_helper_(PeIntegralHelper(basisset)) {
+
+    potentials_ = cppe_state_.get_potentials();
+    cppe_state_.calculate_static_energies_and_fields();
+}
+
+
+std::pair<double, SharedMatrix> PeState::compute_pe_contribution(const SharedMatrix &Dm, CalcType type, bool subtract_scf_density) {
+    // build the electrostatics operator only once!
+    if (!iteration && (type == PeState::CalcType::total)) {
+        V_es_ = std::make_shared<Matrix>("V_es", basisset_->nbf(), basisset_->nbf());
+        for (auto& p : potentials_) {
+            Vector3 site(p.m_x, p.m_y, p.m_z);
+            std::vector<double> moments;
+            for (auto& m : p.get_multipoles()) {
+                m.remove_trace();
+                if (m.m_k > 2) throw std::runtime_error("PE is only implemented up to quadrupoles.");
+                for (auto d : m.get_values()) moments.push_back(d);
+            }
+            V_es_->add(int_helper_.compute_multipole_potential_integrals(site, 2, moments));
+        }
+        cppe_state_.set_es_operator(arma::mat(V_es_->get_pointer(), V_es_->nrow(), V_es_->ncol(), true));
```

Comment:
Would it be possible to add an overload for this function in CPPE that accepts the raw pointer and the sizes and reconstructs the Armadillo matrix internally? It could also take care of saving the matrix to text file. 

I guess this could be a good strategy to avoid Armadillo spilling into Psi4 (or other QC programs, for that matter)

Reply:
Also, why do you need to keep track of the electrostatics operator inside the library?